storyprotocol/protocol-core-v1
# Support Consistent Address Deployment Across Networks

## Description
### Overview
This PR introduces an update to the smart contract deployment process using Foundry and the `CREATE3`. By deploying a standalone `Create3Deployer` contract on networks (e.g., Sepolia), we enable Foundry to execute actual transactions through `Create3Deployer`, and ensuring that contracts are deployed at the same deterministic addresses.

### Key Features
- **Consistent Address Across Chains:** Supports deploying contracts at the same address across all chains by using the same `salt` value derived from a provided `seed`. This ensures uniformity and predictability of contract addresses on various networks.
- **Enhanced Deployment Script:** The deployment script now accepts a `seed` argument. This `seed` is used to generate the `salt` for the Create3 deployment, ensuring that using the same `seed` on different chains will deploy contracts to the same address.

### Motivation
Previously, deploying smart contracts using Foundry with the Create3 library faced verification issues on testnets due to the absence of actual deployment transactions. This PR resolves this issue by ensuring that each deployment involves a real transaction on the blockchain.

### Additional Benefits
This update not only fixes the verification issue but also enhances the flexibility and control we have over the deployment process. It allows for seamless address changes without sacrificing consistency across multiple chains.

### Usage
Deploy contracts by running the deployment script with a specific `seed`:
```bash
forge script script/foundry/deployment/Main.s.sol:Main <seed value e.g. 8> --sig "run(uint256)" --fork-url <rpc url> -vvvv --broadcast  --verify
```

This command will ensure that the same contract addresses are used across different networks, provided the same `seed` is used.


## Test Plan 
- Deploy on Sepolia testnet and verified all contracts.
- Repeat deployment using the same `seed` on different networks and verify the addresses match.

## Related Issue
Closes #131 

## Note
**This PR also include the change in another PR #115  of licensing hook, will rebase after the PR#115 merged.**
